pkg/cleaner/cleaner.go

```diff
@@ -0,0 +1,373 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cleaner
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"time"
+
+	"github.com/redhat-appstudio/pvc-cleaner/pkg/k8s"
+	"github.com/redhat-appstudio/pvc-cleaner/pkg/model"
+	"github.com/redhat-appstudio/pvc-cleaner/pkg/storage"
+	watchapi "k8s.io/apimachinery/pkg/watch"
+
+	pipelinev1 "github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1"
+	"github.com/tektoncd/pipeline/pkg/client/clientset/versioned/typed/pipeline/v1beta1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/cache"
+	watchTool "k8s.io/client-go/tools/watch"
+)
+
+const (
+	DEF_PERIOD      = 3 * time.Minute
+	CLEANUP_TIMEOUT = 600
+)
+
+var isPVCSubPathCleanerRunning = false
+
+type PVCSubPathCleaner struct {
+	pipelineRunApi v1beta1.PipelineRunInterface
+	subPathStorage *storage.PVCSubPathsStorage
+	clientset      *kubernetes.Clientset
+	conf           *k8s.PVCCleanerConfig
+	namespace      string
+}
+
+func NewPVCSubPathCleaner(pipelineRunApi v1beta1.PipelineRunInterface, subPathStorage *storage.PVCSubPathsStorage, clientset *kubernetes.Clientset, conf *k8s.PVCCleanerConfig, namespace string) *PVCSubPathCleaner {
+	return &PVCSubPathCleaner{
+		pipelineRunApi: pipelineRunApi,
+		subPathStorage: subPathStorage,
+		clientset:      clientset,
+		conf:           conf,
+		namespace:      namespace,
+	}
+}
+
+func (cleaner *PVCSubPathCleaner) ScheduleCleanUpSubPathFoldersContent() {
+	for {
+		time.Sleep(DEF_PERIOD)
+
+		log.Println("------------ Schedule cleanup new subpath folders content")
+		if err := cleaner.cleanUpSubPathFoldersContent(); err != nil {
+			log.Print(err)
+		}
+	}
+}
+
+func (cleaner *PVCSubPathCleaner) WatchNewPipelineRuns(storage *storage.PVCSubPathsStorage) {
+	resourceVersion, err := cleaner.conf.GetWatchResourceVersion()
+	if err != nil {
+		log.Fatal(err)
+	}
+	log.Printf("************ Resource version for watch add operation is %s", resourceVersion)
+
+	// Watcher will be closed after some timeout, so we need to re-create watcher https://github.com/kubernetes/client-go/issues/623.
+	// Let's use "NewRetryWatcher" helper for this purpose.
+	retryWatcher, err := watchTool.NewRetryWatcher(resourceVersion, &cache.ListWatch{
+		WatchFunc: func() func(options metav1.ListOptions) (watchapi.Interface, error) {
+			return func(options metav1.ListOptions) (watchapi.Interface, error) {
+				return cleaner.pipelineRunApi.Watch(context.TODO(), metav1.ListOptions{})
+			}
+		}(),
+	})
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	for {
+		event, ok := <-retryWatcher.ResultChan()
+		if !ok {
+			return
```

Comment:
If I understood correctly retry retry/restart mechanism hidden in the watchTool.NewRetryWatcher. That's why we used this package instead of own implementation.